### ISSUE TITLE
Diffusion Probabilistic Modeling for Video Generation

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#184]
# TL; DR

- Autoregressive residual next frame prediction with ConvRNN.
- Continuous Ranked Probability Score (CRPS) to evaluate the coverage (similar to likelihood, but with empirical samples only).

# Motivation

- 3 desiderata of a video generator: (1) Diversity (2) Temporal Consistency (3) Frame Fidelity
- Predictive coding and neural compression: diffuse the residual that is combined with compressed base to yield the final output.
- Autoregressive diffusion modeling with RNN, a straightforward extension of sequential GANs and VAEs.

# Formulation

<img width="1163" alt="image" src="https://user-images.githubusercontent.com/46308407/188346156-6175ba28-cf89-4b74-9341-456380acbcc8.png">

- Factorize the video generation into 2 subtasks: Residual generation (diffusion) and Frame composition (MAF: Masked Autoregressive Flow)

### Residual Diffusion

- The diffusion model (parameterized by $\theta$) takes n-th step residual and the video context to produce the denoised residual.

<img width="558" alt="image" src="https://user-images.githubusercontent.com/46308407/188346598-ed42ffc9-62e3-4c9a-96af-678033252770.png">
<img width="641" alt="image" src="https://user-images.githubusercontent.com/46308407/188346629-9efa0177-cc1e-44e5-985b-2f4e43c5444d.png">

### Frame composition (*data likelihood* in the paper)

<img width="633" alt="image" src="https://user-images.githubusercontent.com/46308407/188346990-fc3b03c7-6ef6-432f-950e-2a5cd5b1d105.png">

- The sparse residuals are temporally *decorrelated*, lowering the task complexity.
- An autoregressive prediction network (parameterized by $\phi$) outputs the context, which is scaled to output the residual.
- The scale parameter is simply fixed to 2 in the implementation. 

<img width="1183" alt="image" src="https://user-images.githubusercontent.com/46308407/188348083-e4f1a392-5bd0-4b73-b735-77f906b0a2ff.png">

# Evaluation

- During training, 2 initial frames are given and the model outputs next 6 frames.
- During inference, the model leverages the initial 4 frames to extend the following 16 frames.

<img width="1179" alt="image" src="https://user-images.githubusercontent.com/46308407/188348097-3ea0d1c0-7fcd-4428-b3ce-65b14818e897.png">

- CRPS: measures the agreement of a CDF with an observation. The CDF is approximated with an empirical CDF with S samples.

<img width="1164" alt="image" src="https://user-images.githubusercontent.com/46308407/188348205-95837851-88be-4689-ba7a-cfa1df6be271.png">
<img width="1184" alt="image" src="https://user-images.githubusercontent.com/46308407/188348411-902a8f3e-d164-457a-a006-fcc85110fbd3.png">
<img width="1175" alt="image" src="https://user-images.githubusercontent.com/46308407/188348423-39e889af-180d-4ddc-8984-015d5ff83aa9.png">
<img width="1164" alt="image" src="https://user-images.githubusercontent.com/46308407/188348433-a6887067-7971-4f99-8eb6-df8c325fec8e.png">

# Inspiration

- New evaluation metric (CRPS) that puts emphasis on true distribution reconsturction (recall).
- Incorporating autoregression into the diffusion framework as in standard sequential VAEs (with ConvRNN).
